confgenerator/config.go

```diff
@@ -140,35 +347,118 @@ type LoggingPipeline struct {
 }
 
 // Ops Agent metrics config.
+type metricsReceiverMap map[string]MetricsReceiver
+type metricsProcessorMap map[string]MetricsProcessor
+type metricsExporterMap map[string]MetricsExporter
 type Metrics struct {
-	Receivers  map[string]*MetricsReceiver  `yaml:"receivers"`
-	Processors map[string]*MetricsProcessor `yaml:"processors"`
-	Exporters  map[string]*MetricsExporter  `yaml:"exporters,omitempty"`
-	Service    *MetricsService              `yaml:"service"`
+	Receivers  metricsReceiverMap  `yaml:"receivers" validate:"dive,keys,startsnotwith=lib:"`
+	Processors metricsProcessorMap `yaml:"processors" validate:"dive,keys,startsnotwith=lib:"`
+	Exporters  metricsExporterMap  `yaml:"exporters,omitempty"`
+	Service    *MetricsService     `yaml:"service"`
 }
 
-type MetricsReceiver struct {
-	configComponent `yaml:",inline"`
+type MetricsReceiver interface {
+	component
+}
 
-	CollectionInterval string `yaml:"collection_interval" validate:"required"` // time.Duration format
+type MetricsReceiverShared struct {
+	CollectionInterval string `yaml:"collection_interval" validate:"required,duration=10s"` // time.Duration format
 }
 
-type MetricsProcessorExcludeMetrics struct {
-	MetricsPattern []string `yaml:"metrics_pattern,flow" validate:"required"`
+var metricsReceiverTypes = &componentTypeRegistry{
+	Subagent: "metrics", Kind: "receiver",
 }
 
-type MetricsProcessor struct {
-	configComponent `yaml:",inline"`
+// Wrapper type to store the unmarshaled YAML value.
+type metricsReceiverWrapper struct {
+	inner interface{}
+}
+
+func (m *metricsReceiverWrapper) UnmarshalYAML(ctx context.Context, unmarshal func(interface{}) error) error {
+	return metricsReceiverTypes.unmarshalComponentYaml(ctx, &m.inner, unmarshal)
+}
+
+func (m *metricsReceiverMap) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	// Unmarshal into a temporary map to capture types.
+	tm := map[string]metricsReceiverWrapper{}
+	if err := unmarshal(&tm); err != nil {
+		return err
+	}
+	// Unwrap the structs.
+	*m = metricsReceiverMap{}
+	for k, r := range tm {
+		if r.inner == nil {
+			return fmt.Errorf("unknown type for receiver %q", k) // FIXME
```

Comment:
Fix FIXME.

Reply:
The `FIXME` was to put a better error message there, so that's not getting fixed in this PR.